### ISSUE TITLE
[JENKINS-41541] Support Agent and environment config at a stage level (currently is only global)

### ISSUE THREAD

**Issue by jenkins-infra-bot** [jenkinsci/blueocean-plugin#3702]
---
<details><summary><i>Originally reported by <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=michaelneale">michaelneale</a>, imported from: <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41541" target="_blank">Support Agent and environment config at a stage level (currently is only global)</a></i></summary>
<i><ul>
<li><b>assignee</b>: <img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/brody.png" title="brody's avatar" /> <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=brody">brody</a>
<li><b>status</b>: Resolved
<li><b>priority</b>: Critical
<li><b>component(s)</b>: blueocean-pipeline-editor-plugin, blueocean-plugin
<li><b>resolution</b>: Not A Defect
<li><b>resolved</b>: 2017-06-06T04:54:03+00:00
<li><b>votes</b>: 0
<li><b>watchers</b>: 4
<li><b>imported</b>: 2025-12-01
</ul></i>
</details>
<details><summary><i>1 attachment</i></summary>

- [image-2017-05-09-14-13-53-982.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/37628/image-2017-05-09-14-13-53-982.png)
  > ![image-2017-05-09-14-13-53-982.png](https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/37628/image-2017-05-09-14-13-53-982.png)
</details>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/kzantow.png" title="kzantow's avatar" /> kzantow:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41541?focusedId=285493&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-285493">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This is actually already done. But it was disabled because it makes no sense to not allow agent config at parallel stages. One of the biggest use cases is building on different platforms, so we need a more generalized solution for this.</p>
</pre>
</details>

</ul>
</details>

<p>This is actually already done. But it was disabled because it makes no sense to not allow agent config at parallel stages. One of the biggest use cases is building on different platforms, so we need a more generalized solution for this.</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41541?focusedId=285494&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-285494">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>This may require some blocking design work cc <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a></p>
</pre>
</details>

</ul>
</details>

<p>This may require some blocking design work cc <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41541?focusedId=285496&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-285496">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>:thumbs up:</p>
</pre>
</details>

</ul>
</details>

<p>:thumbs up:</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41541?focusedId=285502&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-285502">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Stealing this so we can go think about the design</p>
</pre>
</details>

</ul>
</details>

<p>Stealing this so we can go think about the design</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41541?focusedId=298101&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-298101">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>Just bumped the priority of this, as discussed in Sydney: we want to be able to set config per stage as well as setting it per parallel stage. </p>
</pre>
</details>

</ul>
</details>

<p>Just bumped the priority of this, as discussed in Sydney: we want to be able to set config per stage as well as setting it per parallel stage. </p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/michaelneale.png" title="michaelneale's avatar" /> michaelneale:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41541?focusedId=298731&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-298731">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> this is a tricky one - as to do this, we really need to have parallel stages (as per the RFE). So the dependency of this to that is kind of fuzzy. Basically they all need to go together. It makes no sense to have stage config without it supporting it in parallel. </p>

<p> </p>

<p>So we need: </p>
<ul>
	<li>RFE done for parallel stages</li>
	<li>Support for the new parallel stage in visualisation (likely not much impact)</li>
	<li>Treat parallel branches as stage in editor (linked ticket) - a bit of work</li>
	<li>Support for config of stages that are both top level and parallel branch (this requires some design work which I think is outstanding right?)</li>
</ul>


<p> </p>

<p>All kind of as one ball of mud. So when we know we can realistically tackle this, need to get heads together with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a> and <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a> and <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kzantow" class="user-hover" rel="kzantow">kzantow</a></p>
</pre>
</details>

</ul>
</details>

<p><a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=jamesdumay" class="user-hover" rel="jamesdumay">jamesdumay</a> this is a tricky one - as to do this, we really need to have parallel stages (as per the RFE). So the dependency of this to that is kind of fuzzy. Basically they all need to go together. It makes no sense to have stage config without it supporting it in parallel. </p>

<p> </p>

<p>So we need: </p>
<ul>
	<li>RFE done for parallel stages</li>
	<li>Support for the new parallel stage in visualisation (likely not much impact)</li>
	<li>Treat parallel branches as stage in editor (linked ticket) - a bit of work</li>
	<li>Support for config of stages that are both top level and parallel branch (this requires some design work which I think is outstanding right?)</li>
</ul>


<p> </p>

<p>All kind of as one ball of mud. So when we know we can realistically tackle this, need to get heads together with <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=abayer" class="user-hover" rel="abayer">abayer</a> and <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=cliffmeyers" class="user-hover" rel="cliffmeyers">cliffmeyers</a> and <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=kzantow" class="user-hover" rel="kzantow">kzantow</a></p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/jamesdumay.png" title="jamesdumay's avatar" /> jamesdumay:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41541?focusedId=298744&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-298744">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>&gt; Support for config of stages that are both top level and parallel branch (this requires some design work which I think is outstanding right?) </p>

<p>I don't believe so - I think we have designs for Editor config on the stage and parallel level right <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=brody" class="user-hover" rel="brody">brody</a>?</p>
</pre>
</details>

</ul>
</details>

<p>> Support for config of stages that are both top level and parallel branch (this requires some design work which I think is outstanding right?) </p>

<p>I don't believe so - I think we have designs for Editor config on the stage and parallel level right <a href="https://issues.jenkins.io/secure/ViewProfile.jspa?name=brody" class="user-hover" rel="brody">brody</a>?</p>

**Comment by jenkins-infra-bot** [comment]
<details><summary><i><img align="left" width="20" src="https://raw.githubusercontent.com/jenkinsci/artifacts-from-jira-issues/refs/heads/main/avatars/brody.png" title="brody's avatar" /> brody:</i></summary>

<ul>
<li><i>Original <a class="no-jira-link-rewrite" href="https://issues.jenkins.io/browse/JENKINS-41541?focusedId=298747&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-298747">comment link</a></i>
<li><details><summary><i>Raw content of original comment:</i></summary>

<pre>
<p>The handles appear above all columns (with or without parallels). Clicking on the handle shows the column/parallel settings on the right, with Agent &amp; Environment config. <a href="https://invis.io/Q2A110KM9#/211126483_Editor_-_Parallel_Settings" class="external-link" target="_blank" rel="nofollow noopener">https://invis.io/Q2A110KM9#/211126483_Editor_-_Parallel_Settings</a></p>

<p><span class="image-wrap" style=""><img src="https://issues.jenkins.io/secure/attachment/37628/37628_image-2017-05-09-14-13-53-982.png" style="border: 0px solid black" /></span></p>

<p> </p>
</pre>
</details>

</ul>
</details>

<p>The handles appear above all columns (with or without parallels). Clicking on the handle shows the column/parallel settings on the right, with Agent & Environment config. <a href="https://invis.io/Q2A110KM9#/211126483_Editor_-_Parallel_Settings" class="external-link" target="_blank" rel="nofollow noopener">https://invis.io/Q2A110KM9#/211126483_Editor_-_Parallel_Settings</a></p>

<p><span class="image-wrap" style=""><img src="https://raw.githubusercontent.com/jenkinsci/attachments-from-jira-issues-misc/refs/heads/main/attachments/37628/image-2017-05-09-14-13-53-982.png" style="border: 0px solid black" /></span></p>

<p> </p>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `blocks` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-42364%22">JENKINS-42364</a>]</i>

**Comment by jenkins-infra-bot** [comment]
<i>[Original `epic` from Jira: <a href="https://github.com/jenkinsci/blueocean-plugin/issues?q=is%3Aissue%20%22jira_issue_key%3DJENKINS-36291%22">JENKINS-36291</a>]</i>
